### ISSUE TITLE
Support for an operation that serves a file download

### ISSUE THREAD

**Issue by jvieten** [springfox/springfox#562]
Will there be any support for example for something like?

@RequestMapping(value = "/files/{file_name}", method = RequestMethod.GET)
	@ResponseBody
	public FileSystemResource getFile(@PathVariable("file_name") String fileName) {
		return new FileSystemResource(fileName);
	}

**Comment by dilipkrish** [comment]
@jvieten I think it already supports it. So for example if you know what the file type is going to be just add a produces attribute to your request mapping. For e.g. if it produces images you might add something like  `produces = "image/jpeg`. The swagger ui already knows how to render certain media types.

**Comment by jvieten** [comment]
Hi thanks for the reply... I am having some difficulties getting the flow
"make Spring 4  return a File outputstream to the client".


My Resource looks like this:

@RequestMapping(value = "/attachments/{attachmentUid}", method =
RequestMethod.GET, produces = {MediaType.IMAGE_JPEG_VALUE})
@ResponseBody
public FileSystemResource getAttachment(@PathVariable("attachmentId")
final String attachmentId, HttpServletResponse response ) throws
IOException
{
   return  null;
}

But I don't want the FileSystemResource to appear as a model in the
swagger spec.

If I return byte[] the swagger Codegen maps this to a List<byte> .....

Have you any ideas?

Anyway thanks for responding!





2015-01-16 18:42 GMT+01:00 Dilip Krishnan <notifications@github.com>:

**Comment by umbreak** [comment]
Same as the issue before. Using FileSystemResource, even when the producer is "images/jpeg" does not generate a suitable api-docs which can be used to generate the client code.

I tried also using the alternateTypeRules to convert FileSystemResource into something else like ByteArrayProperty or File, but not working either. Some idea how to proceed here?

**Comment by dilipkrish** [comment]
Just returning a `MultipartFile` doesn't work? You should transform `FileSystemResource` to a `MultipartFile` using a rule.

**Comment by umbreak** [comment]
I can confirm that this works. However I couldn't find any documentation for it. So just for people who might be hitting the same wall:

The controller
````java
@RestController
@RequestMapping("v2/api/")
@Api(value = "Resize controller", authorizations = {@Authorization(value="basicAuth")})
public class ResizerController {

    @RequestMapping(method = RequestMethod.POST, path = "resize", produces = {MediaType.IMAGE_JPEG_VALUE})
    public DeferredResult<FileSystemResource> resize(){ ... }
}
````
The configuration:

````java
@EnableSwagger2
public class Application {
    //Any other configuration...
    @Autowired TypeResolver typeResolver;

    @Bean
    public Docket swaggerSettings() {
        List<SecurityScheme> scheme = Lists.newArrayList(new BasicAuth("basicAuth"));
        return new Docket(DocumentationType.SWAGGER_2)
                .alternateTypeRules(
                        newRule(typeResolver.resolve(DeferredResult.class,typeResolver.resolve(FileSystemResource.class)),
                                typeResolver.resolve(MultipartFile.class)))
                .securitySchemes(scheme)
                .select()
                .apis(RequestHandlerSelectors.any())
                .paths(PathSelectors.ant("/v2/api/*"))
                .build()
                .protocols(Sets.newHashSet("http"));
    }
}
````

**Comment by dilipkrish** [comment]
@umbreak thanks!

**Comment by dilipkrish** [comment]
A PR to  [update the docs](https://github.com/springfox/springfox/blob/master/docs/asciidoc/common-problems.adoc) would be much appreciated.

**Comment by vpetit35** [comment]
I don't know if I am missing a point, but I can't get @umbreak solution working.
I keep getting swagger-ui logging "cannot parse JSON/YAML content", no response in the body section (the broken image symbol), and an auto attempt to hit the same path but as a GET this time (even if previous POST call was 200OK).

I am out of idea, I am still not able by any way to get an image as output from a POST in swagger, even though if it is working from postman.

**Comment by dilipkrish** [comment]
@vpetit35 its a bug in [Swagger-ui](https://github.com/swagger-api/swagger-ui)

**Comment by vpetit35** [comment]
Ok I guess this is it [#2226](https://github.com/swagger-api/swagger-ui/issues/2226), that seems resolved in [#2404](https://github.com/swagger-api/swagger-ui/pull/2404).
I pulled the 2.8.0 and it seems fine now.
Thanks

**Comment by ruXlab** [comment]
In my case fix provided by @umbreak didn't work straight away, had to override precedence like in the example provided(in kotlin):

```kotlin
        val typeResolverRules = arrayOf(
            newRule(
                typeResolver.resolve(ResponseEntity::class.java, Any::class.java),
                typeResolver.resolve(MultipartFile::class.java),
                DIRECT_SUBSTITUTION_RULE_ORDER
            )
        )
       
        return Docket(SWAGGER_2)
            . . . . . . .
            .alternateTypeRules(*typeResolverRules)
```

Where `typeResolver` has type of `com.fasterxml.classmate.TypeResolver`
